app/models/hyrax/resource.rb

```diff
@@ -31,6 +31,7 @@ module Hyrax
   #   implementations).
   #
   class Resource < Valkyrie::Resource
+    include Hyrax::Flexibility if Hyrax.config.flexible?
```

Comment:
We would really prefer it if this instead followed the module builder pattern and one did, for example,

```rb
class MyWork < Hyrax::Resource
  include Hyrax::Schema(:my_work, schema_loader: Hyrax::M3SchemaLoader.new(:my_work))
  # or simply…
  include Hyrax::M3Schema(:my_work)
end
```

to define a new work. It really seems like this MR is leaning hard into having everything be magic/automatic, but I don’t think that’s a good thing (it prevents applications from customizing the behaviour). Hyrax should continue supporting multiple means of defining the schema, allow different resources to use different methods, etc… not lock applications into exactly one approach with an opaque `include Hyrax::Flexibility`. This would also potentially obviate the need for `Hyrax.config.flexible?` checks, as it no longer needs to be a global setting.

Additionally, right now it feels like a lot of the schema‐defining code is currently being split between `Hyrax::Flexibility` and `Hyrax::M3SchemaLoader`, and it would be nice for it to all be in one place.

Reply:
I do feel that turning flexible metadata on should be very simple to configure and not require a lot of steps, but that doesn't prevent us from having a more direct structure. Would you agree that moving the include down a level to the children of Hyrax::Resource, maybe moving them to module builder ( the merits of that are suspect to me, but I could be convinced ) and _then_ keeping them behind a configuration flag adds enough customizability?

Reply:
My initial thinking was that turning flexible metadata on should be easy enough that it’s feasible to do in a generator, but maybe doesn’t need to be so easy that it happens automatically with an environment variable. Changing an include from `include Hyrax::Schema(…)` to `include Hyrax::M3Schema(…)` felt reasonable in that context. But this isn’t something I feel particularly strongly about, and I could be persuaded otherwise.

Is it right that the plan is that `Hyrax.config.flexible?` will also eventually be used for other features, like modifying or enabling views? If the plan is for substantial behaviour to eventually be enabled/disabled behind that flag, it makes a lot more sense to group this behaviour with it. But without knowing the eventual scope of the feature flag, it’s hard for me to evaluate whether it feels justified.

I’ll drop some notes about how we’ve handled M3 in Surfliner as another comment, for reference. We don’t have the problem of versioning or live reloading, so I understand that you have some additional needs/constraints. But hopefully it will help us form consensus on how we can best support both approaches.